orttraining/orttraining/eager/opgen/opgen/generator.py

```diff
@@ -424,6 +445,21 @@ def _write_function_body(self, writer: writer.SourceWriter, mapped_func: MappedO
                                     in_place_params[0] = cpp_param.identifier.value
                                     break
 
+                # if no in_place_params found and there is an out input to set
+                # and this is the last onnx op, we set the out to be written to
+                if len(in_place_params) == 0 and set_out_tensor == 1 and onnx_op_index == (len(ctx.ops) - 1):
+                    if not equal_onnxop_found:
```

Comment:
Do you see there being other special cases? 

If so, did you consider structuring this way (which I think would more easily extend to handle other cases):

```python

if is_special_case:
  handle special case
else:
  handle regular case
```

Reply:
I'll reverse the order and now the special case is `cast_output_to_bool`.